keps/sig-api-machinery/2340-Consistent-reads-from-cache/README.md

```diff
@@ -239,139 +229,258 @@ benefits of serving consistent reads from cache. For example, the kubelet LIST
+##### Unit tests
 
-Memory would need to be somehow bound with this approach.
+Unit test with a mock storage backend (instead of an actual etcd) that
+various orderings of progress notify events and "current revision" response
+result in the watch cache serving consistent read requests correctly
 
-#### Rejected Option: Return unpaginated responses to paginated list requests
+- `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/cacher`: `13.06.2023` - `84`
+- `k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd3`: `12.06.2023` - `75`
 
-Both for backward compatibility with older versions of Kubernetes that do not support
-pagination, and for compatibility with api-servers that have the watch cache enabled,
-clients must be able to tolerate unpaginated responses for paginated LIST requests.
+##### Integration tests
 
-The kube-apiserver already switches between serving paginated responses and
-serving unpaginated responses depending on if the watch cache is enabled for the
-types requested.
+##### e2e tests
 
-User cases that have disabled the watch cache will still receive paginated responses.
+Introduce e2e test that run both with etcd progress notify events enabled
+and disable to ensure both configurations work correctly (both with this
+feature enabled and disabled)
 
-Use cases where paginated was being used to reduce load on etcd, but are able to
-tolerate the data volume of the list being returned in a single response, should
-not impacted.
+Benchmark consistent reads from cache against consistent reads from etcd for:
+- list result sizes of 1, 10, ..., 100000
+- object sizes of 5kb, 25kb, 100kb
+- measure latency and throughput
+- document results in this KEP
 
-The most likely problem with this approach is that, because LIST with
-resourceVersion="" is the only way to paginate data from the api-server when the
-watch-cache is enabled (which is the default), clients that need pagination
-(e.g. to avoid receiving too much data in a single request) will be relying on
-LIST resourceVersion="".
+### Graduation Criteria
 
-We are not planning to pursue this option.
+#### Alpha
 
-### What if the watch cache is stale?
+- Feature is implemented behind a feature gate
+- Unpaginated LIST requests is served from watch cache
+- First page of paginated requests is served from watch cache
+- Feature performance is validated via scalability tests
 
-This design requires wait for a watch cache to catch up to the needed revision
-for consistent reads. If the cache doesn't catch up within some time limit we
-either fail the request for have a fallback.
+#### Beta
 
-If the fallback it to forward consistent reads to etcd, a cascading failure
-is likely to occur if caches become stale and a large number of read requests
-are forwarded to etcd.
+- Implement a per-request opt-out from quorum read
```

Comment:
Please link https://github.com/kubernetes/enhancements/pull/1404#discussion_r381528406 here to be able to link to some history

Reply:
SG